L46-core-tls-credential-API.md

```diff
@@ -0,0 +1,390 @@
+1. Inability to choose the maximum/minimum TLS version to be negotiated during the TLS handshake.
+1. No support for certificate revocation.
+
+We propose an API that will meet the following requirements:  
+
+1. Flexible enough to support multiple use cases. Users should be able to write their own reloading or verification logic if the provided ones do not fit their needs.
+1. The credential reloading should be efficient. We shall reload only when needed rather than reloading per TLS connection.
+1. C core API needs to be simple enough, so that most of existing SSL credentials in wrapped languages can migrate to call this new API.
+
+### Related Proposals:
+
+1. [L46: Create a new TLS credential API that supports SPIFFE mutual TLS](https://github.com/grpc/proposal/pull/98) - this was subsumed into this proposal.
+1. [A69: Crl Providers](https://github.com/grpc/proposal/pull/382)
+
+## Proposal
+
+### TLS credentials and TLS Options
+This part of the proposal introduces `TlsCredentialsOptions` and TLS credentials. Users use `TlsChannelCredentialsOptions` and `TlsServerCredentialsOptions` to configure the specific security properties they want, and build the client/server credential by passing in those options.
+
+```c++
+// Base class of configurable options specified by users to configure their
+// certain security features supported in TLS. It is used for experimental
+// purposes for now and it is subject to change.
+class TlsCredentialsOptions {
+ public:
+  // Constructor for base class TlsCredentialsOptions.
+  //
+  // @param certificate_provider the provider which fetches TLS credentials that
+  // will be used in the TLS handshake
+  TlsCredentialsOptions();
+  ~TlsCredentialsOptions();
+
+  // Copy constructor does a deep copy of the underlying pointer. No assignment
+  // permitted
+  TlsCredentialsOptions(const TlsCredentialsOptions& other);
+  TlsCredentialsOptions& operator=(const TlsCredentialsOptions& other) = delete;
+
+  // ---- Setters for member fields ----
+  // Sets the certificate provider used to store root certs and identity certs.
+  void set_certificate_provider(
+      std::shared_ptr<CertificateProviderInterface> certificate_provider);
+  // Watches the updates of root certificates with name |root_cert_name|.
+  // If used in TLS credentials, setting this field is optional for both the
+  // client side and the server side.
+  // If this is not set on the client side, we will use the root certificates
+  // stored in the default system location, since client side must provide root
+  // certificates in TLS(no matter single-side TLS or mutual TLS).
+  // If this is not set on the server side, we will not watch any root
+  // certificate updates, and assume no root certificates needed for the server
+  // (in the one-side TLS scenario, the server is not required to provide root
+  // certs). We don't support default root certs on server side.
```

Comment:
It's the text above - if not explicitly set, we don't try to load system default roots or anything, rather we assume the server doesn't need roots because it does not care to verify clients (i.e. just TLS, not mTLS).

Reply:
I see, thanks for clarifying. I think it's more clear if we just remove this sentence, since it is not adding anything new that we didn't already say in the previous sentence. WDYT?

Reply:
Done